src/Microsoft.AspNet.Routing/RouteData.cs

```diff
@@ -32,24 +34,55 @@ public RouteData(RouteData other)
                 throw new ArgumentNullException(nameof(other));
             }
 
-            DataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
             Routers = new List<IRouter>(other.Routers);
-            Values = new RouteValueDictionary(other.Values);
+            
+            // Perf: Avoid allocating DataTokens and RouteValues unless we need to make a copy.
+            if (other._dataTokens != null)
+            {
+                _dataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
```

Comment:
I went ahead and tried this, and was only barely measurably better, but a whole lot more complex to test and code. Leaving this change as-is.